### PR TITLE
Add a transaction to put tokens back into a subnetwork's treasury

### PR DESCRIPTION
**Pull request by Vagabond**

### CODE REVIEW

**Comment by PaulVMo** [comment]
Should the nonce be checked in `is_valid` similar to what is done for a payment?

File: src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    0.
+
+-spec fee_payer(txn_subnetwork_fund_v1(), blockchain_ledger_v1:ledger()) ->
+    libp2p_crypto:pubkey_bin() | undefined.
+fee_payer(Txn, _Ledger) ->
+    payer(Txn).
+
+-spec is_valid(txn_subnetwork_fund_v1(), blockchain:blockchain()) ->
+    ok | {error, atom()} | {error, {atom(), any()}}.
+is_valid(Txn, Chain) ->
```
Comment (andymck) [inline src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl:98]: missing nonce check here

Comment (Vagabond) [reply src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl:98]: its checked in the absorb, which is how its done in, for example, payment_v2?

Comment (Vagabond) [reply src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl:98]: ah, no, you're right, let me do it in is_valid as well.

File: src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    to_json/2
+]).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-type txn_subnetwork_fund_v1() :: #blockchain_txn_subnetwork_fund_v1_pb{}.
+
+-export_type([txn_subnetwork_fund_v1/0]).
```
Comment (jeffgrunewald) [inline src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl:42]:
```suggestion
-export_type([txn_subnetwork_fund/0]).
```
fwiw it looks like we haven't been including the version suffix on the exported erlang type

**Review by jeffgrunewald — APPROVED** [review]

File: src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        true ->
+            case Amount > 0 of
+                false ->
+                    {error, zero_amount};
+                true ->
+                    %% check the token type is valid
+                    Subnetworks = blockchain_ledger_v1:subnetworks_v1(Ledger),
+                    case maps:is_key(TokenType, Subnetworks) of
+                        false ->
+                            {error, unknown_token_type};
```
Comment (vihu) [inline src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl:120]:
```suggestion
                            {error, {unknown_token_type, TokenType}};
```

File: src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                                    TxnNonce = ?MODULE:nonce(Txn),
+                                    LedgerEntryNonce = blockchain_ledger_entry_v2:nonce(PayerEntry),
+                                    case TxnNonce =:= LedgerEntryNonce + 1 of
+                                        false ->
+                                            {error, {bad_nonce, {subnetwork_fund_v1, TxnNonce, LedgerEntryNonce}}};
+                                        true ->
+                                            PayerTTBalance = blockchain_ledger_entry_v2:balance(PayerEntry, TokenType),
+                                            case PayerTTBalance >= Amount of
+                                                false ->
+                                                    {error, insufficient_balance};
```
Comment (vihu) [inline src/transactions/v1/blockchain_txn_subnetwork_fund_v1.erl:134]:
```suggestion
                                                    {error, {insufficient_balance, PayerTTBalance, Amount}};
```

**Comment by Vagabond** [comment]
All comments addressed.

File: rebar.lock [hunk 5]
```diff
@@ -64,7 +64,7 @@
  {<<"hackney">>,{pkg,<<"hackney">>,<<"1.18.1">>},0},
  {<<"helium_proto">>,
   {git,"https://github.com/helium/proto.git",
-       {ref,"b09456889e406889f4fa1ea325b9503678dc8447"}},
+       {ref,"afb829b33d52de10626067ae87cb1108b3937e9c"}},
```
Comment (jeffgrunewald) [inline rebar.lock:67]: this sha doesn't seem to exist so the build is broken; otherwise changes look good

**Comment by Vagabond** [comment]
> Should the nonce be checked in `is_valid` similar to what is done for a payment?

This has been addressed

**Pull request merged**